### PR TITLE
SH - CAS / Login token login

### PR DESCRIPTION
**Pull request by stevenhammerton**
Update the CAS login flow to allow the HS to do all the interaction with CAS and use the HS token login

### CODE REVIEW

**Comment by stevenhammerton** [comment]
Will need to go out at same time as https://github.com/matrix-org/matrix-react-sdk/pull/28 and https://github.com/vector-im/vector-web/pull/335

File: lib/http-api.js [hunk 1]
```diff
@@ -296,6 +296,15 @@ module.exports.MatrixHttpApi.prototype = {
         return this._request(callback, method, fullUri, queryParams, data);
     },
 
+    redirect: function(path, queryParams, prefix) {
+        var queryString = "";
+        if (queryParams) {
+            queryString = "?" + utils.encodeParams(queryParams);
+        }
+        var fullUri = this.opts.baseUrl + prefix + path + queryString;
+        window.location.href = fullUri;
```
Comment (Kegsay) [inline lib/http-api.js]: The JS-SDK can run in any JS environment, including those without a browser. Please delegate responsibility for doing the redirect to React SDK.

File: lib/client.js [hunk 2]
```diff
@@ -1722,27 +1722,23 @@ MatrixClient.prototype.loginWithSAML2 = function(relayState, callback) {
 };
 
 /**
- * @param {module:client.callback} callback Optional.
- * @return {module:client.Promise} Resolves: TODO
- * @return {module:http-api.MatrixError} Rejects: with an error response.
+ * @param {string} redirectUrl URL to which to be redirect to after
+ * HS authenticates with CAS and issues login token
+ * Will redirect to homeserver to complete CAS login
```
Comment (Kegsay) [inline lib/client.js]: Wha-? I sort of understand but the grammar is weird. Perhaps:

```
@param {string} redirectUrl The URL to redirect to after the HS authenticates with CAS.
```

Taking into account my other comment, you may want to also do:

```
@return {string} The HS URL to hit to begin the CAS login process.
```

**Comment by Kegsay** [comment]
LGTM other than this.

**Comment by Kegsay** [comment]
LGTM

**Pull request merged**